dash/llmq/llmq.go

```diff
@@ -0,0 +1,256 @@
+package llmq
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"sort"
+
+	bls "github.com/dashpay/bls-signatures/go-bindings"
+
+	"github.com/tendermint/tendermint/crypto"
+	"github.com/tendermint/tendermint/crypto/bls12381"
+)
+
+var (
+	errThresholdInvalid      = errors.New("threshold must be greater than 0")
+	errKeySharesNotGenerated = errors.New("to initialize shares you must generate the keys")
+)
+
+type optionFunc func(c *llmqConfig)
+
+// Data contains pre generated keys/shares/signatures for the participants of LLMQ network
+type Data struct {
+	Threshold        int
+	ProTxHashes      []crypto.ProTxHash
+	PrivKeys         []crypto.PrivKey
+	PubKeys          []crypto.PubKey
+	PrivKeyShares    []crypto.PrivKey
+	PubKeyShares     []crypto.PubKey
+	ThresholdPrivKey crypto.PrivKey
+	ThresholdPubKey  crypto.PubKey
+	ThresholdSig     []byte
+	QuorumHash       crypto.QuorumHash
+}
+
+// blsLLMQData is an intermediate structure that contains the BLS keys/shares/signatures
+type blsLLMQData struct {
+	proTxHashes []crypto.ProTxHash
+	sks         []*bls.PrivateKey
+	pks         []*bls.PublicKey
+	skShares    []*bls.PrivateKey
+	pkShares    []*bls.PublicKey
+}
+
+// MustGenerate generates long-living master node quorum, but panics if a got error
+func MustGenerate(proTxHashes []crypto.ProTxHash, opts ...optionFunc) *Data {
+	data, err := Generate(proTxHashes, opts...)
+	if err != nil {
+		panic(err)
+	}
+	return data
+}
+
+// Generate generates long living master node quorum for a list of pro-tx-hashes
+// to be able to override default values, need to provide option functions
+func Generate(proTxHashes []crypto.ProTxHash, opts ...optionFunc) (*Data, error) {
+	conf := llmqConfig{
+		proTxHashes: bls12381.ReverseProTxHashes(proTxHashes),
+		threshold:   len(proTxHashes)*2/3 + 1,
+		seedReader:  crypto.CReader(),
+	}
+	_, err := crypto.CReader().Read(conf.hash[:])
+	if err != nil {
+		return nil, err
+	}
+	for _, opt := range opts {
+		opt(&conf)
+	}
+	err = conf.validate()
+	if err != nil {
+		return nil, err
+	}
+
+	// sorting makes this easier
+	sort.Sort(crypto.SortProTxHash(conf.proTxHashes))
+
+	ld, err := initLLMQData(
+		conf,
+		initKeys(conf.seedReader),
+		initShares(),
+		// as this is not used in production, we can add this test
+		initValidation(),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return newLLMQDataFromBLSData(ld, conf.threshold), nil
+}
+
+// WithSeed sets a seed generator with passed seed-source
+func WithSeed(seedSource int64) func(c *llmqConfig) {
+	return func(c *llmqConfig) {
+		if seedSource > 0 {
+			c.seedReader = rand.New(rand.NewSource(seedSource))
+		}
+	}
+}
+
+// WithThreshold sets a threshold number of allowed members for
+// a recovery a threshold public key / signature or private key
+func WithThreshold(threshold int) func(c *llmqConfig) {
```

Comment:
To consider: use higher abstraction level based on quorum types instead of (or in addition to) specifying threshold directly, see [func  (vals *ValidatorSet) QuorumTypeThresholdCount()](https://github.com/dashevo/tenderdash/blob/ee269bd6899e94511647a7bc4fcb4baba72d9248/types/validator_set.go#L548) to understand what I mean.